Issues while using s3 select_object_content

Hey guys, I'm new to using moto and I keep encountering an error (Sorry if its a duplicate). I saw select_object_content was highly experimental so I wanted to know if this is my error or maybe it isn't supported at the moment. I realized if None is in the Body while using the select_object_content, the test fails (the object actually goes into the s3 bucket)

Also I have another issue while using queries similar to `select * from s3object[*].names[*]`. It seems that might not be supported, but I guess I'll create another issue to talk more in-depth on it. Thank you!
Moto version: 4.1.10 (Installed using pip)
boto3: 1.26.138
botocore: 1.29.138

 Here's a code snippet:
```
import json

import boto3
from moto import mock_s3


@mock_s3
def test_s3_select():

    s3_client = boto3.client("s3", region_name="us-east-1")
    s3_client.create_bucket(Bucket="mybucket")


    s3_client.put_object(
        Bucket="mybucket",
        Key="tet.json",
        Body=json.dumps([{
            "name": "Mary",
            "date": "2020-02-23T00:00:00",
            "city": "Chicago",
            "kids": None  # it fails ( "kids" : 3 This would work)
        },
        ])
    )

    result = s3_client.select_object_content(
        Bucket="mybucket",
        Key="tet.json",
        ExpressionType="SQL",
        Expression="select * from s3object s",
        InputSerialization={'JSON': {'Type': 'DOCUMENT'}},
        OutputSerialization={"JSON": {}},
    )

    print(result)
```

Error traceback:

```
    @mock_s3
    def test_s3_select():
    
        s3_client = boto3.client("s3", region_name="us-east-1")
        s3_client.create_bucket(Bucket="mybucket")
    
    
        s3_client.put_object(
            Bucket="mybucket",
            Key="tet.json",
            Body=json.dumps([{
                "name": "Janelyn M",
                "date": "2020-02-23T00:00:00",
                "city": "Chicago",
                "kids": None
            },
            ])
        )
    
>       result = s3_client.select_object_content(
            Bucket="mybucket",
            Key="tet.json",
            ExpressionType="SQL",
            Expression="select * from s3object s",
            InputSerialization={'JSON': {'Type': 'DOCUMENT'}},
            OutputSerialization={"JSON": {}},
        )

example.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.10/site-packages/botocore/client.py:530: in _api_call
    return self._make_api_call(operation_name, kwargs)
venv/lib/python3.10/site-packages/botocore/client.py:947: in _make_api_call
    http, parsed_response = self._make_request(
venv/lib/python3.10/site-packages/botocore/client.py:970: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
venv/lib/python3.10/site-packages/botocore/endpoint.py:119: in make_request
    return self._send_request(request_dict, operation_model)
venv/lib/python3.10/site-packages/botocore/endpoint.py:202: in _send_request
    while self._needs_retry(
venv/lib/python3.10/site-packages/botocore/endpoint.py:354: in _needs_retry
    responses = self._event_emitter.emit(
venv/lib/python3.10/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.10/site-packages/botocore/retryhandler.py:207: in __call__
    if self._checker(**checker_kwargs):
venv/lib/python3.10/site-packages/botocore/retryhandler.py:284: in __call__
    should_retry = self._should_retry(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:307: in _should_retry
    return self._checker(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:363: in __call__
    checker_response = checker(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:247: in __call__
    return self._check_caught_exception(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:416: in _check_caught_exception
    raise caught_exception
venv/lib/python3.10/site-packages/botocore/endpoint.py:278: in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
venv/lib/python3.10/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.10/site-packages/moto/core/botocore_stubber.py:61: in __call__
    status, headers, body = response_callback(
venv/lib/python3.10/site-packages/moto/s3/responses.py:250: in ambiguous_response
    return self.key_response(request, full_url, headers)
venv/lib/python3.10/site-packages/moto/utilities/aws_headers.py:74: in _wrapper
    response = f(*args, **kwargs)
venv/lib/python3.10/site-packages/moto/s3/responses.py:1154: in key_response
    response = self._key_response(request, full_url, self.headers)
venv/lib/python3.10/site-packages/moto/s3/responses.py:1274: in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
venv/lib/python3.10/site-packages/moto/s3/responses.py:2217: in _key_response_post
    results = self.backend.select_object_content(
venv/lib/python3.10/site-packages/moto/s3/models.py:2489: in select_object_content
    return [
venv/lib/python3.10/site-packages/moto/s3/models.py:2490: in <listcomp>
    json.dumps(x, indent=None, separators=(",", ":")).encode("utf-8")
/usr/lib/python3.10/json/__init__.py:238: in dumps
    **kw).encode(obj)
/usr/lib/python3.10/json/encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
/usr/lib/python3.10/json/encoder.py:257: in iterencode
    return _iterencode(o, 0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x7f30dfc93880>, o = <None>

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type Variable is not JSON serializable

/usr/lib/python3.10/json/encoder.py:179: TypeError
=================================================================================================== short test summary info ===================================================================================================
FAILED example.py::test_s3_select - TypeError: Object of type Variable is not JSON serializable
```

Awesome! I'll test it out. Thanks alot for your hard work @bblommers 
Regarding the other issue, here is an example:
```
import json

import boto3
from moto import mock_s3


@mock_s3
def test_s3_select():

    s3_client = boto3.client("s3", region_name="us-east-1")
    s3_client.create_bucket(Bucket="mybucket")

    s3_client.put_object(
        Bucket="mybucket",
        Key="tet.json",
        Body=json.dumps([{
            "staff":[
                {
                    "name": "Janelyn M",
                    "date": "2020-02-23T00:00:00",
                    "city": "Chicago",
                    "kids": 2
                },
                {
                    "name": "Stacy P",
                    "date": "2020-02-28T00:00:00",
                    "city": "Seattle",
                    "kids": 1
                }
            ],
            "country": "USA"
        }])
    )

    result = s3_client.select_object_content(
        Bucket="mybucket",
        Key="tet.json",
        ExpressionType="SQL",
        Expression="select * from s3object[*].staff[*] s",
        InputSerialization={'JSON': {'Type': 'DOCUMENT'}},
        OutputSerialization={"JSON": {}},
    )
```
It raises a key error due to [*] PS, using this query expression `select * from s3object s` also didn't work for this (different error). Here's the KeyError:
```
_______________________________________________________________________________________________________ test_s3_select ________________________________________________________________________________________________________

    @mock_s3
    def test_s3_select():
    
        s3_client = boto3.client("s3", region_name="us-east-1")
        s3_client.create_bucket(Bucket="mybucket")
    
        s3_client.put_object(
            Bucket="mybucket",
            Key="tet.json",
            Body=json.dumps([{
                "staff":[
                    {
                        "name": "Janelyn M",
                        "date": "2020-02-23T00:00:00",
                        "city": "Chicago",
                        "kids": 2
                    },
                    {
                        "name": "Stacy P",
                        "date": "2020-02-28T00:00:00",
                        "city": "Seattle",
                        "kids": 1
                    }
                ],
                "country": "USA"
            }])
        )
    
>       result = s3_client.select_object_content(
            Bucket="mybucket",
            Key="tet.json",
            ExpressionType="SQL",
            Expression="select * from s3object[*].staff[*] s",
            InputSerialization={'JSON': {'Type': 'DOCUMENT'}},
            OutputSerialization={"JSON": {}},
        )

example.py:35: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.10/site-packages/botocore/client.py:530: in _api_call
    return self._make_api_call(operation_name, kwargs)
venv/lib/python3.10/site-packages/botocore/client.py:947: in _make_api_call
    http, parsed_response = self._make_request(
venv/lib/python3.10/site-packages/botocore/client.py:970: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
venv/lib/python3.10/site-packages/botocore/endpoint.py:119: in make_request
    return self._send_request(request_dict, operation_model)
venv/lib/python3.10/site-packages/botocore/endpoint.py:202: in _send_request
    while self._needs_retry(
venv/lib/python3.10/site-packages/botocore/endpoint.py:354: in _needs_retry
    responses = self._event_emitter.emit(
venv/lib/python3.10/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.10/site-packages/botocore/retryhandler.py:207: in __call__
    if self._checker(**checker_kwargs):
venv/lib/python3.10/site-packages/botocore/retryhandler.py:284: in __call__
    should_retry = self._should_retry(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:307: in _should_retry
    return self._checker(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:363: in __call__
    checker_response = checker(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:247: in __call__
    return self._check_caught_exception(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:416: in _check_caught_exception
    raise caught_exception
venv/lib/python3.10/site-packages/botocore/endpoint.py:278: in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
venv/lib/python3.10/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.10/site-packages/moto/core/botocore_stubber.py:61: in __call__
    status, headers, body = response_callback(
venv/lib/python3.10/site-packages/moto/s3/responses.py:250: in ambiguous_response
    return self.key_response(request, full_url, headers)
venv/lib/python3.10/site-packages/moto/utilities/aws_headers.py:74: in _wrapper
    response = f(*args, **kwargs)
venv/lib/python3.10/site-packages/moto/s3/responses.py:1154: in key_response
    response = self._key_response(request, full_url, self.headers)
venv/lib/python3.10/site-packages/moto/s3/responses.py:1274: in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
venv/lib/python3.10/site-packages/moto/s3/responses.py:2217: in _key_response_post
    results = self.backend.select_object_content(
venv/lib/python3.10/site-packages/moto/s3/models.py:2491: in select_object_content
    for x in parse_query(query_input, select_query)
venv/lib/python3.10/site-packages/moto/s3/select_object_content.py:9: in parse_query
    return S3SelectParser(source_data={"s3object": text_input}).parse(query)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <py_partiql_parser._internal.parser.S3SelectParser object at 0x7f5a84ec3be0>, query = 'select * from s3object[*].staff[*] s', parameters = None

    def parse(self, query: str, parameters=None) -> List[Dict[str, Any]]:
        query = query.replace("\n", " ")
        clauses = re.split("SELECT | FROM | WHERE ", query, flags=re.IGNORECASE)
        # First clause is whatever comes in front of SELECT - which should be nothing
        _ = clauses[0]
        # FROM
        from_clauses = FromParser().parse(clauses[2])
>       source_data = self.documents[list(from_clauses.values())[0].lower()]
E       KeyError: '[*]'

venv/lib/python3.10/site-packages/py_partiql_parser/_internal/parser.py:36: KeyError
=================================================================================================== short test summary info ===================================================================================================
FAILED example.py::test_s3_select - KeyError: '[*]'
```

Here's the other error i noticed:

```
    @mock_s3
    def test_s3_select():
    
        s3_client = boto3.client("s3", region_name="us-east-1")
        s3_client.create_bucket(Bucket="mybucket")
    
        s3_client.put_object(
            Bucket="mybucket",
            Key="tet.json",
            Body=json.dumps([{
                "staff":[
                    {
                        "name": "Janelyn M",
                        "date": "2020-02-23T00:00:00",
                        "city": "Chicago",
                        "kids": 2
                    },
                    {
                        "name": "Stacy P",
                        "date": "2020-02-28T00:00:00",
                        "city": "Seattle",
                        "kids": 1
                    }
                ],
                "country": "USA"
            }])
        )
    
>       result = s3_client.select_object_content(
            Bucket="mybucket",
            Key="tet.json",
            ExpressionType="SQL",
            Expression="select * from s3object s",
            InputSerialization={'JSON': {'Type': 'DOCUMENT'}},
            OutputSerialization={"JSON": {}},
        )

example.py:35: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
venv/lib/python3.10/site-packages/botocore/client.py:530: in _api_call
    return self._make_api_call(operation_name, kwargs)
venv/lib/python3.10/site-packages/botocore/client.py:947: in _make_api_call
    http, parsed_response = self._make_request(
venv/lib/python3.10/site-packages/botocore/client.py:970: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
venv/lib/python3.10/site-packages/botocore/endpoint.py:119: in make_request
    return self._send_request(request_dict, operation_model)
venv/lib/python3.10/site-packages/botocore/endpoint.py:202: in _send_request
    while self._needs_retry(
venv/lib/python3.10/site-packages/botocore/endpoint.py:354: in _needs_retry
    responses = self._event_emitter.emit(
venv/lib/python3.10/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.10/site-packages/botocore/retryhandler.py:207: in __call__
    if self._checker(**checker_kwargs):
venv/lib/python3.10/site-packages/botocore/retryhandler.py:284: in __call__
    should_retry = self._should_retry(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:307: in _should_retry
    return self._checker(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:363: in __call__
    checker_response = checker(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:247: in __call__
    return self._check_caught_exception(
venv/lib/python3.10/site-packages/botocore/retryhandler.py:416: in _check_caught_exception
    raise caught_exception
venv/lib/python3.10/site-packages/botocore/endpoint.py:278: in _do_get_response
    responses = self._event_emitter.emit(event_name, request=request)
venv/lib/python3.10/site-packages/botocore/hooks.py:412: in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:256: in emit
    return self._emit(event_name, kwargs)
venv/lib/python3.10/site-packages/botocore/hooks.py:239: in _emit
    response = handler(**kwargs)
venv/lib/python3.10/site-packages/moto/core/botocore_stubber.py:61: in __call__
    status, headers, body = response_callback(
venv/lib/python3.10/site-packages/moto/s3/responses.py:250: in ambiguous_response
    return self.key_response(request, full_url, headers)
venv/lib/python3.10/site-packages/moto/utilities/aws_headers.py:74: in _wrapper
    response = f(*args, **kwargs)
venv/lib/python3.10/site-packages/moto/s3/responses.py:1154: in key_response
    response = self._key_response(request, full_url, self.headers)
venv/lib/python3.10/site-packages/moto/s3/responses.py:1274: in _key_response
    return self._key_response_post(request, body, bucket_name, query, key_name)
venv/lib/python3.10/site-packages/moto/s3/responses.py:2217: in _key_response_post
    results = self.backend.select_object_content(
venv/lib/python3.10/site-packages/moto/s3/models.py:2489: in select_object_content
    return [
venv/lib/python3.10/site-packages/moto/s3/models.py:2490: in <listcomp>
    json.dumps(x, indent=None, separators=(",", ":")).encode("utf-8")
/usr/lib/python3.10/json/__init__.py:238: in dumps
    **kw).encode(obj)
/usr/lib/python3.10/json/encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
/usr/lib/python3.10/json/encoder.py:257: in iterencode
    return _iterencode(o, 0)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x7f0433512560>, o = {'name': 'Janelyn M', 'date': '2020-02-23T00:00:00', 'city': 'Chicago', 'kids': 2}

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type CaseInsensitiveDict is not JSON serializable

/usr/lib/python3.10/json/encoder.py:179: TypeError
=================================================================================================== short test summary info ===================================================================================================
FAILED example.py::test_s3_select - TypeError: Object of type CaseInsensitiveDict is not JSON serializable
```

--- a/setup.cfg
+++ b/setup.cfg
@@ -52,7 +52,7 @@ all =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.3.1
+    py-partiql-parser==0.3.3
     aws-xray-sdk!=0.96,>=0.93
     setuptools
 server =
@@ -66,7 +66,7 @@ server =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.3.1
+    py-partiql-parser==0.3.3
     aws-xray-sdk!=0.96,>=0.93
     setuptools
     flask!=2.2.0,!=2.2.1
@@ -100,7 +100,7 @@ cloudformation =
     openapi-spec-validator>=0.2.8
     pyparsing>=3.0.7
     jsondiff>=1.1.2
-    py-partiql-parser==0.3.1
+    py-partiql-parser==0.3.3
     aws-xray-sdk!=0.96,>=0.93
     setuptools
 cloudfront =
@@ -123,10 +123,10 @@ dms =
 ds = sshpubkeys>=3.1.0
 dynamodb =
     docker>=3.0.0
-    py-partiql-parser==0.3.1
+    py-partiql-parser==0.3.3
 dynamodbstreams =
     docker>=3.0.0
-    py-partiql-parser==0.3.1
+    py-partiql-parser==0.3.3
 ebs = sshpubkeys>=3.1.0
 ec2 = sshpubkeys>=3.1.0
 ec2instanceconnect =
@@ -183,7 +183,7 @@ route53 =
 route53resolver = sshpubkeys>=3.1.0
 s3 =
     PyYAML>=5.1
-    py-partiql-parser==0.3.1
+    py-partiql-parser==0.3.3
 s3control =
 sagemaker =
 sdb =

--- a/tests/test_s3/test_s3_select.py
+++ b/tests/test_s3/test_s3_select.py
@@ -8,6 +8,15 @@ from uuid import uuid4
 
 SIMPLE_JSON = {"a1": "b1", "a2": "b2", "a3": None}
 SIMPLE_JSON2 = {"a1": "b2", "a3": "b3"}
+EXTENSIVE_JSON = [
+    {
+        "staff": [
+            {"name": "Janelyn M", "city": "Chicago", "kids": 2},
+            {"name": "Stacy P", "city": "Seattle", "kids": 1},
+        ],
+        "country": "USA",
+    }
+]
 SIMPLE_LIST = [SIMPLE_JSON, SIMPLE_JSON2]
 SIMPLE_CSV = """a,b,c
 e,r,f
@@ -30,11 +39,17 @@ class TestS3Select(TestCase):
         self.client.put_object(
             Bucket=self.bucket_name, Key="simple_csv", Body=SIMPLE_CSV
         )
+        self.client.put_object(
+            Bucket=self.bucket_name,
+            Key="extensive.json",
+            Body=json.dumps(EXTENSIVE_JSON),
+        )
 
     def tearDown(self) -> None:
         self.client.delete_object(Bucket=self.bucket_name, Key="list.json")
         self.client.delete_object(Bucket=self.bucket_name, Key="simple.json")
         self.client.delete_object(Bucket=self.bucket_name, Key="simple_csv")
+        self.client.delete_object(Bucket=self.bucket_name, Key="extensive.json")
         self.client.delete_bucket(Bucket=self.bucket_name)
 
     def test_query_all(self):
@@ -119,3 +134,31 @@ class TestS3Select(TestCase):
         )
         result = list(x["Payload"])
         result.should.contain({"Records": {"Payload": b'{"_1":3},'}})
+
+    def test_extensive_json__select_list(self):
+        x = self.client.select_object_content(
+            Bucket=self.bucket_name,
+            Key="extensive.json",
+            Expression="select * from s3object[*].staff[*] s",
+            ExpressionType="SQL",
+            InputSerialization={"JSON": {"Type": "DOCUMENT"}},
+            OutputSerialization={"JSON": {"RecordDelimiter": ","}},
+        )
+        result = list(x["Payload"])
+        assert {"Records": {"Payload": b"{},"}} in result
+
+    def test_extensive_json__select_all(self):
+        x = self.client.select_object_content(
+            Bucket=self.bucket_name,
+            Key="extensive.json",
+            Expression="select * from s3object s",
+            ExpressionType="SQL",
+            InputSerialization={"JSON": {"Type": "DOCUMENT"}},
+            OutputSerialization={"JSON": {"RecordDelimiter": ","}},
+        )
+        result = list(x["Payload"])
+        assert {
+            "Records": {
+                "Payload": b'{"_1":[{"staff":[{"name":"Janelyn M","city":"Chicago","kids":2},{"name":"Stacy P","city":"Seattle","kids":1}],"country":"USA"}]},'
+            }
+        } in result

